channel_acceptor.go

```diff
@@ -0,0 +1,134 @@
+	wg             sync.WaitGroup
+}
+
+// NewChannelAcceptor creates a new channel acceptor with the given lnd client.
+func NewChannelAcceptor(lightning lndclient.LightningClient) *ChannelAcceptor {
+	return &ChannelAcceptor{
+		lightning:     lightning,
+		expectedChans: make(map[[32]byte]*order.Bid),
+		quit:          make(chan struct{}),
+	}
+}
+
+// Start starts the channel acceptor and subscribes to receive all incoming
+// channel events of lnd.
+func (s *ChannelAcceptor) Start(errChan chan error) error {
+	s.errChan = errChan
+
+	ctxc := context.Background()
+	ctxc, s.acceptorCancel = context.WithCancel(ctxc)
+
+	rpcErrChan, err := s.lightning.ChannelAcceptor(
+		ctxc, order.DefaultBatchStepTimeout, s.acceptChannel,
+	)
+	if err != nil {
+		return err
+	}
+
+	s.wg.Add(1)
+	go s.subscribe(rpcErrChan)
+
+	return nil
+}
+
+// subscribe subscribes to errors coming from the RPC error channel and forwards
+// them to our main error channel.
+func (s *ChannelAcceptor) subscribe(rpcErrChan chan error) {
+	defer s.wg.Done()
+
+	for {
+		select {
+		case err := <-rpcErrChan:
+			select {
+			case s.errChan <- err:
+			case <-s.quit:
+			}
+
+		case <-s.quit:
+			s.acceptorCancel()
+
+			return
+		}
+	}
+}
+
+// Stop shuts down the channel acceptor.
+func (s *ChannelAcceptor) Stop() {
+	s.acceptorCancel()
+	close(s.quit)
+
+	s.wg.Wait()
+}
+
+// NotifyShimCreated is a function that should be called whenever a funding shim
+// is created for a bid order where we expect an incoming channel at any moment.
+func (s *ChannelAcceptor) ShimRegistered(bid *order.Bid, pid [32]byte) {
+	s.expectedChansMtx.Lock()
+	defer s.expectedChansMtx.Unlock()
+
+	// We always just add the current expectation we have since an
+	// order might be go through this state multiple times with different
+	// pending channel IDs.
+	// TODO(guggero): Clean up after failure or completion.
+	s.expectedChans[pid] = bid
+}
+
+// acceptChannel is the callback that is invoked each time a new incoming
+// channel message is received in lnd. We inspect it here and if it corresponds
+// to a pending channel ID that we have an expectation for, we check whether the
+// self chan balance (=push amount) is correct.
+func (s *ChannelAcceptor) acceptChannel(_ context.Context,
+	req *lndclient.AcceptorRequest) (*lndclient.AcceptorResponse, error) {
+
+	s.expectedChansMtx.Lock()
+	defer s.expectedChansMtx.Unlock()
+
+	expectedChanBid, ok := s.expectedChans[req.PendingChanID]
+
+	// It's not a channel we've registered within the funding manager so we
+	// just accept it to not interfere with the normal node operation.
+	if !ok {
+		return &lndclient.AcceptorResponse{Accept: true}, nil
+	}
+
+	// The push amount in the acceptor request is in milli sats, we need to
+	// convert it first.
+	pushAmtSat := lnwire.MilliSatoshi(req.PushAmt).ToSatoshis()
+
+	// Push amount must be exactly what we expect. Otherwise the asker could
+	// be trying to cheat.
+	if expectedChanBid.SelfChanBalance != pushAmtSat {
```

Comment:
The upside of doing it here though as that we don't need to update `lnd` each time there's some additional constraint we want to assert.

Reply:
Yeah, it allows us to move forward with this now. Can always "upgrade" this once we have everything we need in the funding shim itself.